File: tvr-decoder/src/main/java/com/willcurrie/tlv/BerTlv.java
package com.willcurrie.tlv;

import java.nio.BufferUnderflowException;
import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public abstract class BerTlv {

    private final Tag tag;

    public BerTlv(Tag tag) {
        this.tag = tag;
    }

    public static BerTlv newInstance(Tag tag, byte[] value) {
        return new PrimitiveBerTlv(tag, value);
    }

    public static BerTlv newInstance(Tag tag, String hexString) {
        return new PrimitiveBerTlv(tag, ISOUtil.hex2byte(hexString));
    }

    public static BerTlv newInstance(Tag tag, int value) {
        if (value > 255) {
            throw new IllegalArgumentException("Value greater than 255 must be encode in a byte array");
        }
        return new PrimitiveBerTlv(tag, new byte[] {(byte) value});
    }

    public static BerTlv newInstance(Tag tag, List<BerTlv> value) {
        return new ConstructedBerTlv(tag, value);
    }

    public static BerTlv newInstance(Tag tag, BerTlv tlv1, BerTlv tlv2) {
        return new ConstructedBerTlv(tag, Arrays.asList(tlv1, tlv2));
    }

    public static BerTlv parse(byte[] data) {
        List<BerTlv> tlvs = parseList(ByteBuffer.wrap(data), true);
        return tlvs.get(0);
    }

    public static BerTlv parseAsPrimitiveTag(byte[] data) {
        List<BerTlv> tlvs = parseList(ByteBuffer.wrap(data), false);
        return tlvs.get(0);
    }

    public static List<BerTlv> parseList(byte[] data, boolean parseConstructedTags) {
        return parseList(ByteBuffer.wrap(data), parseConstructedTags);
    }

    private static List<BerTlv> parseList(ByteBuffer data, boolean parseConstructedTags) {
        List<BerTlv> tlvs = new ArrayList<BerTlv>();

        while (data.hasRemaining()) {
            Tag tag = Tag.parse(data);
            try {
                int length = parseLength(data);

                byte[] value = new byte[length];
                data.get(value);
                if (tag.isConstructed() && parseConstructedTags) {
                    try {
                        tlvs.add(newInstance(tag, parseList(value, parseConstructedTags)));
                    } catch (Exception e) {
                        tlvs.add(newInstance(tag, value));
                    }
                } else {
                    tlvs.add(newInstance(tag, value));
                }
            } catch (Exception e) {
                throw new RuntimeException("Failed parsing " + tag + "," + (e.getMessage() == null ? e.getClass().getSimpleName() : e.getMessage()), e);
            }
        }
        return tlvs;
    }

    private static int parseLength(ByteBuffer data) {
        int lengthByte = data.get();
        int dataLength = 0;
        if ((lengthByte & 0x80) == 0x80) {
            int numberOfBytesToEncodeLength = (lengthByte & 0x7F);
            while (numberOfBytesToEncodeLength > 0) {
                dataLength += data.get() & 0xFF;

                if (numberOfBytesToEncodeLength > 1) {
                    dataLength *= 256;
                }
                numberOfBytesToEncodeLength--;
            }
        } else {
            dataLength = lengthByte;
        }
        return dataLength;
    }

    public static BerTlv findTlv(List<BerTlv> tlvs, Tag tag) {
        for (BerTlv tlv : tlvs) {
            if (tlv.getTag().equals(tag)) {
                return tlv;
            }
        }
        return null;
    }

    public Tag getTag() {
        return tag;
    }

    public byte[] toBinary() {
        byte[] value = getValue();
        byte[] encodedTag = tag.getBytes();
        byte[] encodedLength = getLength(value);

        ByteBuffer b = ByteBuffer.allocate(encodedTag.length + encodedLength.length + value.length);
        b.put(encodedTag);
        b.put(encodedLength);
        b.put(value);
        b.flip();
        return b.array();
    }

    public String toHexString() {
        return ISOUtil.hexString(toBinary());
    }

    public String getValueAsHexString() {
        return ISOUtil.hexString(getValue());
    }

    public int getLengthInBytesOfEncodedLength() {
    	return getLength(getValue()).length;
    }
    
    public abstract BerTlv findTlv(Tag tag);

    public abstract List<BerTlv> findTlvs(Tag tag);

    public abstract byte[] getValue();

    public abstract List<BerTlv> getChildren();

    private byte[] getLength(byte[] value) {
        byte[] length = null;
        if (value == null) {
            return new byte[] {(byte) 0x00};
        }
        if (value.length <= 127) {
            length = new byte[] {(byte) value.length};
        } else {
            int wanted = value.length;
            int expected = 0x100;
            int needed = 1;
            while (!(wanted < expected)) {
                needed++;
                expected = expected << 8;
                if (expected == 0) {
                    // just to be sure
                    throw new IllegalArgumentException();
                }
            }
            length = new byte[needed + 1];
            length[0] = (byte) (0x80 | needed);
            for (int i = 1; i < length.length; i++) {
                length[length.length - i] = (byte) ((wanted >> (8 * (i - 1))) & 0xFF);
            }

        }
        return length;
    }

}

File: src/main/java/io/github/binaryfoo/tlv/BerTlv.kt
package io.github.binaryfoo.tlv

import java.nio.ByteBuffer
import java.util.ArrayList
import java.util.Arrays
import kotlin.collections.firstOrNull

/**
 * Model data elements encoded using the Basic Encoding Rules: http://en.wikipedia.org/wiki/X.690#BER_encoding.
 */
public abstract class BerTlv(public val tag: Tag) {

    public fun toBinary(): ByteArray {
        val value = getValue()
        val encodedTag = tag.bytes
        val encodedLength = getLength(value)

        val b = ByteBuffer.allocate(encodedTag.size + encodedLength.size + value.size)
        b.put(encodedTag.toByteArray())
        b.put(encodedLength)
        b.put(value)
        b.flip()
        return b.array()
    }

    /**
     * The whole object (the T, L and V components) as a hex string.
     */
    public fun toHexString(): String = ISOUtil.hexString(toBinary())

    /**
     * The value of V in TLV as a hex string.
     */
    public val valueAsHexString: String
    get() = ISOUtil.hexString(getValue())

    /**
     * The number of bytes used to encode the L (length) in TLV.
     * Eg 1 byte might be used to encode a length of 12, whilst at least 2 bytes would be used for a length of 300.
     */
    public val lengthInBytesOfEncodedLength: Int
    get() = getLength(getValue()).size

    /**
     * The value of L (length) in TLV. Length in bytes of the value.
     */
    public val length: Int
    get() = getValue().size

    /**
     * Skip the tag and length bytes.
     */
    public val startIndexOfValue: Int
    get() = tag.bytes.size + lengthInBytesOfEncodedLength

    public abstract fun findTlv(tag: Tag): BerTlv?

    public abstract fun findTlvs(tag: Tag): List<BerTlv>

    /**
     * The value of V in TLV as a byte array.
     */
    public abstract fun getValue(): ByteArray

    /**
     * For a constructed TLV the child elements that make up the V. For a primitive, an empty list.
     */
    public abstract fun getChildren(): List<BerTlv>

    private fun getLength(value: ByteArray?): ByteArray {
        val length: ByteArray
        if (value == null) {
            return byteArrayOf(0.toByte())
        }
        if (value.size <= 0x7F) {
            length = byteArrayOf(value.size.toByte())
        } else {
            val wanted = value.size
            var expected = 256
            var needed = 1
            while (wanted >= expected) {
                needed++
                expected = expected shl 8
                if (expected == 0) {
                    // just to be sure
                    throw IllegalArgumentException()
                }
            }
            length = ByteArray(needed + 1)
            length[0] = (0x80 or needed).toByte()
            for (i in 1..length.size - 1) {
                length[length.size - i] = ((wanted shr (8 * (i - 1))) and 255).toByte()
            }

        }
        return length
    }

    companion object {

        @JvmStatic public fun newInstance(tag: Tag, value: ByteArray): BerTlv {
            return PrimitiveBerTlv(tag, value)
        }

        @JvmStatic public fun newInstance(tag: Tag, hexString: String): BerTlv {
            return PrimitiveBerTlv(tag, ISOUtil.hex2byte(hexString))
        }

        @JvmStatic public fun newInstance(tag: Tag, value: Int): BerTlv {
            if (value > 255) {
                throw IllegalArgumentException("Value greater than 255 must be encoded in a byte array")
            }
            return PrimitiveBerTlv(tag, byteArrayOf(value.toByte()))
        }

        @JvmStatic public fun newInstance(tag: Tag, value: List<BerTlv>): BerTlv {
            return ConstructedBerTlv(tag, value)
        }

        @JvmStatic public fun newInstance(tag: Tag, tlv1: BerTlv, tlv2: BerTlv): BerTlv {
            return ConstructedBerTlv(tag, Arrays.asList<BerTlv>(tlv1, tlv2))
        }

        @JvmStatic public fun parse(data: ByteArray): BerTlv {
            return parseList(ByteBuffer.wrap(data), true)[0]
        }

        @JvmStatic public fun parseAsPrimitiveTag(data: ByteArray): BerTlv {
            return parseList(ByteBuffer.wrap(data), false)[0]
        }

        @JvmStatic public fun parseList(data: ByteArray, parseConstructedTags: Boolean): List<BerTlv> {
            return parseList(ByteBuffer.wrap(data), parseConstructedTags)
        }

        @JvmStatic public fun parseList(data: ByteArray, parseConstructedTags: Boolean, recognitionMode: TagRecognitionMode): List<BerTlv> {
            return parseList(ByteBuffer.wrap(data), parseConstructedTags, recognitionMode)
        }

        private fun parseList(data: ByteBuffer, parseConstructedTags: Boolean, recognitionMode: TagRecognitionMode = CompliantTagMode): List<BerTlv> {
            val tlvs = ArrayList<BerTlv>()

            while (data.hasRemaining()) {
                val tag = Tag.parse(data, recognitionMode)
                if (isPaddingByte(tag)) {
                    continue
                }
                try {
                    val length = parseLength(data)
                    val value = readUpToLength(data, length)
                    if (tag.constructed && parseConstructedTags) {
                        try {
                            tlvs.add(newInstance(tag, parseList(value, true, recognitionMode)))
                        } catch (e: Exception) {
                            tlvs.add(newInstance(tag, value))
                        }

                    } else {
                        tlvs.add(newInstance(tag, value))
                    }
                } catch (e: Exception) {
                    throw TlvParseException(tlvs, "Failed parsing TLV with tag $tag: " + e.message, e)
                }

            }
            return tlvs
        }

        private fun readUpToLength(data: ByteBuffer, length: Int): ByteArray {
            val value = ByteArray(if (length > data.remaining()) data.remaining() else length)
            data.get(value)
            return value
        }

        // Specification Update No. 69, 2009, Padding of BER-TLV Encoded Constructed Data Objects
        private fun isPaddingByte(tag: Tag): Boolean {
            return tag.bytes.size == 1 && tag.bytes[0] == 0.toByte()
        }

        private fun parseLength(data: ByteBuffer): Int {
            val firstByte = data.get().toInt()
            var length = 0
            if ((firstByte and 0x80) == 0x80) {
                var numberOfBytesToEncodeLength = (firstByte and 0x7F)
                for (i in 1..numberOfBytesToEncodeLength) {
                    if (!data.hasRemaining()) {
                        throw IllegalArgumentException("Bad length: expected to read $numberOfBytesToEncodeLength (0x${firstByte.toByte().toHexString()}) bytes. Only have ${i-1}.")
                    }
                    length += (data.get().toInt() and 0xFF)
                    if (i != numberOfBytesToEncodeLength) {
                        length *= 256
                    }
                    if (length < 0) {
                        throw IllegalArgumentException("Bad length: $length < 0. Read $i of $numberOfBytesToEncodeLength (0x${firstByte.toByte().toHexString()}) bytes used to encode length of TLV.")
                    }
                }
            } else {
                length = firstByte
            }
            return length
        }

        @JvmStatic public fun findTlv(tlvs: List<BerTlv>, tag: Tag): BerTlv? = tlvs.firstOrNull { it.tag == tag }
    }

}
